File: output/xCodeEval_8000/C++/C/b2d0ed53da3b843a0b67aece5f924b78.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>

typedef long long int ll;

ll max(ll a, ll b) {
    return (a > b) ? a : b;
}

void solve() {
    int n, k;
    scanf("%d %d", &n, &k);

    ll sums[n];
    int arrs[n][k];
    for (int i = 0; i < n; i++) {
        int t;
        scanf("%d", &t);
        for (int j = 0; j < t; j++) {
            scanf("%d", &arrs[i][j]);
        }
        if (t > k) {
            t = k;
        }
        sums[i] = 0;
        for (int j = 0; j < t; j++) {
            sums[i] += arrs[i][j];
        }
    }

    ll dps[13][k+1];
    memset(dps, 0, sizeof(dps));

    for (int i = 0; i < n; i++) {
        ll ans = 0;
        int carr[k];
        memcpy(carr, arrs[i], sizeof(carr));

        if (i + 1 == n) {
            for (int j = 0; j <= k; j++) {
                ans = max(ans, dps[0][j]);
            }
            ll sum = 0;
            for (int j = 0; j < k && j < sizeof(carr) / sizeof(carr[0]); j++) {
                sum += carr[j];
                ans = max(ans, dps[0][k-j-1] + sum);
            }
        } else {
            int m = (i + n) / 2;
            memcpy(dps[1], dps[0], sizeof(dps[0]));
            for (int j = i; j < m; j++) {
                ll v = sums[j];
                int k1 = sizeof(arrs[j]) / sizeof(arrs[j][0]);
                for (int l = k; l >= k1; l--) {
                    dps[1][l] = max(dps[1][l], dps[1][l-k1] + v);
                }
            }
            ans = max(ans, rec(m, n, 1));
            memcpy(dps[1], dps[0], sizeof(dps[0]));
            for (int j = m; j < n; j++) {
                ll v = sums[j];
                int k1 = sizeof(arrs[j]) / sizeof(arrs[j][0]);
                for (int l = k; l >= k1; l--) {
                    dps[1][l] = max(dps[1][l], dps[1][l-k1] + v);
                }
            }
            ans = max(ans, rec(i, m, 1));
        }

        printf("%lld\n", ans);
    }
}

int main() {
    solve();
    return 0;
}

File: output/GPT3.5-Turbo/xCodeEval_4000/C++/Go/b2d0ed53da3b843a0b67aece5f924b78.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"sort"
	"strconv"
)

func max(a, b int64) int64 {
	if a > b {
		return a
	}
	return b
}

func sum(arr []int64) int64 {
	var s int64
	for _, num := range arr {
		s += num
	}
	return s
}

func knapsack(arrs [][]int64, k int) int64 {
	n := len(arrs)
	sums := make([]int64, n)
	for i := 0; i < n; i++ {
		sums[i] = sum(arrs[i])
	}

	dps := make([][]int64, 13)
	for i := range dps {
		dps[i] = make([]int64, k+1)
	}

	var rec func(l, r, dpi int) int64
	rec = func(l, r, dpi int) int64 {
		ans := int64(0)
		carr := arrs[l]
		dp := dps[dpi]

		if l+1 == r {
			ans = max(ans, dp[k])
			sum := int64(0)
			for i := 0; i < min(k, len(carr)); i++ {
				sum += carr[i]
				ans = max(ans, dp[k-i-1]+sum)
			}
		} else {
			m := (l + r) / 2
			copy(dps[dpi+1], dps[dpi])
			for i := l; i < m; i++ {
				v := sums[i]
				k1 := len(arrs[i])
				for j := k; j >= k1; j-- {
					dps[dpi+1][j] = max(dps[dpi+1][j], dps[dpi+1][j-k1]+v)
				}
			}
			ans = max(ans, rec(m, r, dpi+1))

			copy(dps[dpi+1], dps[dpi])
			for i := m; i < r; i++ {
				v := sums[i]
				k1 := len(arrs[i])
				for j := k; j >= k1; j-- {
					dps[dpi+1][j] = max(dps[dpi+1][j], dps[dpi+1][j-k1]+v)
				}
			}
			ans = max(ans, rec(l, m, dpi+1))
		}

		return ans
	}

	return rec(0, n, 0)
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func solve(r io.Reader, w io.Writer) {
	scanner := bufio.NewScanner(r)
	scanner.Split(bufio.ScanWords)

	var n, k int
	scanner.Scan()
	n, _ = strconv.Atoi(scanner.Text())
	scanner.Scan()
	k, _ = strconv.Atoi(scanner.Text())

	arrs := make([][]int64, n)
	for i := 0; i < n; i++ {
		scanner.Scan()
		t, _ := strconv.Atoi(scanner.Text())
		arrs[i] = make([]int64, t)
		for j := 0; j < t; j++ {
			scanner.Scan()
			arrs[i][j], _ = strconv.ParseInt(scanner.Text(), 10, 64)
		}
		if t > k {
			arrs[i] = arrs[i][:k]
		}
	}

	sort.Slice(arrs, func(i, j int) bool {
		return sum(arrs[i]) > sum(arrs[j])
	})

	ans := knapsack(arrs, k)
	fmt.Fprintln(w, ans)
}

func main() {
	solve(os.Stdin, os.Stdout)
}